### ISSUE TITLE
`dipy==1.9.0` is causing our macOS GitHub Actions runners to stall on unrelated `torch` tests

### ISSUE THREAD

**Issue by joshuacwnewton** [dipy/dipy#3125]
## Description

This is a bit of a weird one. In a [routine maintenance PR](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4332) where we're upgrading the versions of Python packages installed into our virtual environment, we noticed that some [tests were timing out](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4332#discussion_r1525678494) for macOS runners. After a litany of debugging, we narrowed it down to `dipy==1.9.0`. (Holding [all other packages](https://www.diffchecker.com/e6kEdkO9/) at their same versions, `dipy==1.8.0` [passes](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8290169407/job/22687795876#step:4:38) but `dipy==1.9.0` [times out](https://github.com/spinalcordtoolbox/spinalcordtoolbox/actions/runs/8299368372/job/22714903875#step:4:39).)

The tests that stall appear to be ones related to `torch`-based deep leaning pipelines. For example, we have a deep learning-based registration model called `algo='dl'` that uses the packages [`voxelmorph`](https://github.com/voxelmorph/voxelmorph)/`neurite`/`pystrum`. 

<details>
<summary> pipdeptree output for voxelmorph </summary>

```
voxelmorph==0.2
├── h5py [required: Any, installed: 3.10.0]
│   └── numpy [required: >=1.17.3, installed: 1.26.4]
├── neurite [required: >=0.2, installed: 0.2]
│   ├── matplotlib [required: Any, installed: 3.8.3]
│   │   ├── contourpy [required: >=1.0.1, installed: 1.2.0]
│   │   │   └── numpy [required: >=1.20,<2.0, installed: 1.26.4]
│   │   ├── cycler [required: >=0.10, installed: 0.12.1]
│   │   ├── fonttools [required: >=4.22.0, installed: 4.49.0]
│   │   ├── importlib-resources [required: >=3.2.0, installed: 6.3.0]
│   │   │   └── zipp [required: >=3.1.0, installed: 3.18.1]
│   │   ├── kiwisolver [required: >=1.3.1, installed: 1.4.5]
│   │   ├── numpy [required: >=1.21,<2, installed: 1.26.4]
│   │   ├── packaging [required: >=20.0, installed: 24.0]
│   │   ├── pillow [required: >=8, installed: 10.2.0]
│   │   ├── pyparsing [required: >=2.3.1, installed: 3.1.2]
│   │   └── python-dateutil [required: >=2.7, installed: 2.9.0.post0]
│   │       └── six [required: >=1.5, installed: 1.16.0]
│   ├── nibabel [required: Any, installed: 5.2.1]
│   │   ├── numpy [required: >=1.20, installed: 1.26.4]
│   │   └── packaging [required: >=17, installed: 24.0]
│   ├── numpy [required: >=1.17, installed: 1.26.4]
│   ├── packaging [required: Any, installed: 24.0]
│   ├── pystrum [required: >=0.2, installed: 0.4]
│   │   ├── matplotlib [required: Any, installed: 3.8.3]
│   │   │   ├── contourpy [required: >=1.0.1, installed: 1.2.0]
│   │   │   │   └── numpy [required: >=1.20,<2.0, installed: 1.26.4]
│   │   │   ├── cycler [required: >=0.10, installed: 0.12.1]
│   │   │   ├── fonttools [required: >=4.22.0, installed: 4.49.0]
│   │   │   ├── importlib-resources [required: >=3.2.0, installed: 6.3.0]
│   │   │   │   └── zipp [required: >=3.1.0, installed: 3.18.1]
│   │   │   ├── kiwisolver [required: >=1.3.1, installed: 1.4.5]
│   │   │   ├── numpy [required: >=1.21,<2, installed: 1.26.4]
│   │   │   ├── packaging [required: >=20.0, installed: 24.0]
│   │   │   ├── pillow [required: >=8, installed: 10.2.0]
│   │   │   ├── pyparsing [required: >=2.3.1, installed: 3.1.2]
│   │   │   └── python-dateutil [required: >=2.7, installed: 2.9.0.post0]
│   │   │       └── six [required: >=1.5, installed: 1.16.0]
│   │   ├── numpy [required: Any, installed: 1.26.4]
│   │   ├── scipy [required: Any, installed: 1.12.0]
│   │   │   └── numpy [required: >=1.22.4,<1.29.0, installed: 1.26.4]
│   │   └── six [required: Any, installed: 1.16.0]
│   ├── scikit-learn [required: Any, installed: 1.4.1.post1]
│   │   ├── joblib [required: >=1.2.0, installed: 1.3.2]
│   │   ├── numpy [required: >=1.19.5,<2.0, installed: 1.26.4]
│   │   ├── scipy [required: >=1.6.0, installed: 1.12.0]
│   │   │   └── numpy [required: >=1.22.4,<1.29.0, installed: 1.26.4]
│   │   └── threadpoolctl [required: >=2.0.0, installed: 3.3.0]
│   ├── scipy [required: Any, installed: 1.12.0]
│   │   └── numpy [required: >=1.22.4,<1.29.0, installed: 1.26.4]
│   ├── six [required: Any, installed: 1.16.0]
│   └── tqdm [required: Any, installed: 4.66.2]
├── nibabel [required: Any, installed: 5.2.1]
│   ├── numpy [required: >=1.20, installed: 1.26.4]
│   └── packaging [required: >=17, installed: 24.0]
├── numpy [required: Any, installed: 1.26.4]
├── packaging [required: Any, installed: 24.0]
├── scikit-image [required: Any, installed: 0.22.0]
│   ├── imageio [required: >=2.27, installed: 2.34.0]
│   │   ├── numpy [required: Any, installed: 1.26.4]
│   │   └── pillow [required: >=8.3.2, installed: 10.2.0]
│   ├── lazy-loader [required: >=0.3, installed: 0.3]
│   ├── networkx [required: >=2.8, installed: 3.2.1]
│   ├── numpy [required: >=1.22, installed: 1.26.4]
│   ├── packaging [required: >=21, installed: 24.0]
│   ├── pillow [required: >=9.0.1, installed: 10.2.0]
│   ├── scipy [required: >=1.8, installed: 1.12.0]
│   │   └── numpy [required: >=1.22.4,<1.29.0, installed: 1.26.4]
│   └── tifffile [required: >=2022.8.12, installed: 2024.2.12]
│       └── numpy [required: Any, installed: 1.26.4]
└── scipy [required: Any, installed: 1.12.0]
    └── numpy [required: >=1.22.4,<1.29.0, installed: 1.26.4]
```

</details>

However, I've found that our usage of `dipy` is entirely independent of this test. In fact, if I uninstall `dipy` locally and comment out our `dipy` imports, the test is still able to be run just fine without error on my local Ubuntu machine. So, I'm at a bit of a loss as to why the `dipy` upgrade would cause this test (as well as other non-voxelmorph tests) to time out, and only for our `macos` GHA runners.

I'd be happy to do more debugging + narrow things down to a more reproducible example. But, before I do that, I wanted to report this issue upstream, just in case my general description of the issue rings a bell for a relevant change that dipy may have made between 1.8.0 and 1.9.0. 

(Skimming https://github.com/dipy/dipy/releases/tag/1.9.0, the only macOS-related change I see is https://github.com/dipy/dipy/pull/3061/, but I'm not entirely sure if that's related.)

## Way to reproduce
[If reporting a bug, please include the following important information:]
- [ ] Code example
- [ ] Relevant images (if any)
- [ ] Operating system and version (run `python -c "import platform; print(platform.platform())"`)
- [ ] Python version (run `python -c "import sys; print('Python', sys.version)"`)
- [ ] dipy version (run `python -c "import dipy; print(dipy.__version__)"`)
- [ ] dependency version (numpy, scipy, nibabel, h5py, cvxpy, fury)
    * import numpy; print("NumPy", numpy.__version__)
    * import scipy; print("SciPy", scipy.__version__)
    * import nibabel; print("Nibabel", nibabel.__version__)
    * import h5py; print("H5py", h5py.__version__)
    * import cvxpy; print("Cvxpy", cvxpy.__version__)
    * import fury; print("fury", fury.__version__)

> [!NOTE]
> To be added by @joshuacwnewton.

**Comment by skoudoro** [comment]
Hi @joshuacwnewton,

I confirm, The difference between 1.8.0 and 1.9.0 on macOS is OPENMP:
since 1.9.0, we build our package with openmp flag which was not the case before.

Not sure what is the issue. So maybe you are encountering concurrency, or openmp conflict, or something like that with pytorch. I will experiment next week

**Comment by joshuacwnewton** [comment]
Thank you for the confirmation! I'll try some experimentation on our end, too. :)

**Comment by joshuacwnewton** [comment]
The OpenMP mention reminded me of a similar issue we had with an upstream change to `onnxruntime`:

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/2930
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3033
- https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/3433

tl;dr: `onnxruntime` required OpenMP to be installed for versions 1.5.x and 1.6.x. They then removed this requirement in 1.7.0 due to feedback: 

- https://github.com/microsoft/onnxruntime/pull/6610

Going off of what we did previously to fix this (install openmp in CI), I tried this on our test PR to see if it would prevent the runner from timing out. But, `libomp` is already installed:

```
Run brew install libomp
Warning: libomp 17.0.6 is already installed and up-to-date.
```

I will keep investigating, but my hunch is that our package will just have to specific `dipy!=1.9.0` in the short term.

**Comment by larsoner** [comment]
Perhaps `threadpoolctl` or `OMP_NUM_THREADS=1` could help? See for example https://github.com/scikit-learn/scikit-learn/issues/28625#issuecomment-2007600405 and related comments about multiple packages with openmp not playing nicely with one another.

**Comment by joshuacwnewton** [comment]
Hullo again! My apologies for the delay here. I'm back to debugging this issue in: 

- https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4644

----

I'm finding now that instead of hanging, the macOS runner now hits a segfault. This seems to be the case regardless of whether `OMP_NUM_THREADS=1` is set in the context of the test.

<details>
<summary> Full traceback </summary>


```
Fatal Python error: Segmentation fault

Thread 0xFatal Python error: Segmentation fault

00007000090e3000 (most recent call first):
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/threading.py", line 316 in wait
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/threading.py", line 581 in wait
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/tqdm/_monitor.py", line 60 in run
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/threading.py", line 980 in _bootstrap_inner
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/threading.py", line 937 in _bootstrap

Thread 0x0000000113478600 (most recent call first):
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/dipy/denoise/nlmeans.py", line 57 in nlmeans
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/spinalcordtoolbox/scripts/sct_denoising_onlm.py", line 155 in main
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/testing/cli/test_cli_sct_denoising_onlm.py", line 16 in test_sct_denoising_onlm_no_checks
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/python.py", line 159 in pytest_pyfunc_call
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/python.py", line 1627 in runtest
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/runner.py", line 174 in pytest_runtest_call
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/runner.py", line 242 in <lambda>
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/runner.py", line 341 in from_call
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/runner.py", line 241 in call_and_report
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/runner.py", line 132 in runtestprotocol
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/runner.py", line 113 in pytest_runtest_protocol
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/main.py", line 362 in pytest_runtestloop
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/main.py", line 337 in _main
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/main.py", line 283 in wrap_session
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_callers.py", line 103 in _multicall
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/pluggy/_hooks.py", line 513 in __call__
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/config/__init__.py", line 175 in main
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/site-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/bin/pytest", line 8 in <module>
```
</details>

Interestingly, setting both [`OMP_NUM_THREADS=1` + `OPENBLAS_NUM_THREADS=1`](https://github.com/scikit-learn/scikit-learn/issues/28625#issuecomment-2009581604) in the context of the `nlmeans` call seems to fix it, as does using `threadpoolctl.threading_limits`.

However, confusingly, after the problematic test passes, the _next_ test seems to immediately fail, even though that second test doesn't use `dipy` or `torch`. 

```
testing/cli/test_cli_sct_denoising_onlm.py::test_sct_denoising_onlm_no_checks PASSED [ 65%]
Fatal Python error: Segmentation fault

Thread 0x0000700009bd1000 (most recent call first):
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/threading.py", line 316 in wait
  File "/Users/runner/work/spinalcordtoolbox/spinalcordtoolbox/python/envs/venv_sct/lib/python3.9/threading.py", line 581 in wait
./.ci.sh: line 49:  4715 Segmentation fault: 11  pytest testing/api testing/cli
testing/cli/test_cli_sct_detect_pmj.py::test_sct_detect_pmj_check_euclidean_distance_against_groundtruth
```

I wasn't sure if somehow the thread-specific issues were "leaking" so to speak, once the context of the first test was exited. To test the theory, I then tried setting the environment variables at the call to `pytest` (to cover the entire process), and the tests all pass.

----

In summary, I think our solution may end up being:

- Use `threadpoolctl` to wrap individual calls to the problematic API (for users, etc.)
- Use the environment variables in the context of `pytest`'s cfg (e.g. via [`pytest-env`](https://pypi.org/project/pytest-env/) to make sure this doesn't trip up our CI.

If you can think of a better solution, or have insight into the weird pytest behavior, feel free to let us know though!

**Comment by joshuacwnewton** [comment]
Note that given the possible need for both OpenMP + OpenBLAS limits, maybe following scikit-learn's lead here with https://github.com/scikit-learn/scikit-learn/pull/28692 could be helpful to avoid the need for workarounds?
